File: chart/src/main/java/cn/jingzhuan/lib/chart/data/AbstractDataSet.java
package cn.jingzhuan.lib.chart.data;

import android.graphics.Color;
import android.graphics.Rect;

import cn.jingzhuan.lib.chart.Viewport;
import cn.jingzhuan.lib.chart.component.AxisY;
import cn.jingzhuan.lib.chart.component.AxisY.AxisDependency;
import cn.jingzhuan.lib.chart.utils.RequestDataType;

import java.util.ArrayList;
import java.util.List;

/**
 * Abstract DataSet
 * <p>
 * Created by Donglua on 17/7/20.
 */

public abstract class AbstractDataSet<T extends Value> extends AbstractVisible implements IDataSet {

  protected float mViewportYMin = Float.MAX_VALUE;
  protected float mViewportYMax = -Float.MAX_VALUE;
  protected float minValueOffsetPercent = 0F;
  protected float maxValueOffsetPercent = 0F;
  protected float startXOffset = 0f;
  protected float endXOffset = 0f;
  private int mAxisDependency = AxisY.DEPENDENCY_LEFT;
  private int mColor = Color.GRAY;
  private int maxVisibleEntryCount = 500;
  private int minVisibleEntryCount = 15;
  private int defaultVisibleEntryCount = -1;
  private boolean isHighlightedVerticalEnable = false;
  private boolean isHighlightedHorizontalEnable = false;

  private boolean enable = true;

  private int minValueCount = -1;

  private int drawIndex = -1;

  private String tag;

  private DataFormatter formatter; // 数据格式化器

  public AbstractDataSet() {
  }

  public AbstractDataSet(String tag) {
    this.tag = tag;
  }

  @Override
  public void calcMinMax(Viewport viewport) {}
  @Override
  public void calcMinMax(Viewport viewport, Rect content, float max, float mix) {
    calcMinMax(viewport);
  }

  public abstract List<T> getValues();

  public abstract void setValues(List<T> values);

  public abstract boolean addEntry(T e);

  public abstract boolean removeEntry(T e);

  public abstract int getEntryIndex(T e);

  public abstract T getEntryForIndex(int index);

  @AxisDependency
  public int getAxisDependency() {
    return mAxisDependency;
  }

  public void setAxisDependency(@AxisDependency int mAxisDependency) {
    this.mAxisDependency = mAxisDependency;
  }

  public float getViewportYMin() {
    return mViewportYMin;
  }

  public float getViewportYMax() {
    return mViewportYMax;
  }

  public int getColor() {
    return mColor;
  }

  public void setColor(int barColor) {
    this.mColor = barColor;
  }

  public boolean isHighlightedVerticalEnable() {
    return isHighlightedVerticalEnable;
  }

  public void setHighlightedVerticalEnable(boolean highlightedVerticalEnable) {
    isHighlightedVerticalEnable = highlightedVerticalEnable;
  }

  public boolean isHighlightedHorizontalEnable() {
    return isHighlightedHorizontalEnable;
  }

  public void setHighlightedHorizontalEnable(boolean highlightedHorizontalEnable) {
    isHighlightedHorizontalEnable = highlightedHorizontalEnable;
  }

  public List<T> getVisiblePoints(Viewport viewport) {
    ArrayList<T> allValue = new ArrayList<>(getValues());
    int listSize = allValue.size();

    int from = Math.round(viewport.left * listSize);
    int to = Math.round(viewport.right * listSize);

    if (listSize <= minVisibleEntryCount) {
      return allValue;
    }

    return safeSubList(allValue, from, to);
  }

  private List<T> safeSubList(List<T> list, int from, int to) {
    if (list == null || list.isEmpty()) return list;
    int size = list.size();
    int safeFrom =  (from >= size) ? 0 : from;
    int safeTo = Math.min(to, size);
    return list.subList(safeFrom, safeTo);
  }

  public float getVisibleRange(Viewport viewport) {
    return (viewport.right - viewport.left) * getEntryCount();
  }

  @Override
  public void setMaxVisibleEntryCount(int maxVisibleEntryCount) {
    this.maxVisibleEntryCount = maxVisibleEntryCount;
  }

  public int getMaxVisibleEntryCount() {
    return maxVisibleEntryCount;
  }

  @Override
  public void setMinVisibleEntryCount(int minVisibleEntryCount) {
    this.minVisibleEntryCount = minVisibleEntryCount;
  }

  public int getMinVisibleEntryCount() {
    return minVisibleEntryCount;
  }

  @Override
  public void setDefaultVisibleEntryCount(int defaultVisibleEntryCount) {
    this.defaultVisibleEntryCount = defaultVisibleEntryCount;
  }

  public int getDefaultVisibleEntryCount() {
    return defaultVisibleEntryCount;
  }

  @Override
  public boolean isEnable() {
    return enable;
  }

  public void setEnable(boolean enable) {
    this.enable = enable;
    this.setVisible(enable);
  }

  @Override
  public int getMinValueCount() {
    return minValueCount;
  }

  @Override
  public void setMinValueCount(int minValueCount) {
    this.minValueCount = minValueCount;
  }

  public int getStartIndexOffset() {
    int startIndex = 0;
//    if (minValueCount > 0 && getValues() != null && getValues().size() > 0) {
//      startIndex = getEntryCount() - getValues().size();
//    }
    return startIndex;
  }

  public int getDrawIndex() {
    return drawIndex;
  }

  public void setDrawIndex(int drawIndex) {
    this.drawIndex = drawIndex;
  }

  public String getTag() {
    return tag;
  }

  public void setTag(String tag) {
    this.tag = tag;
  }

  public DataFormatter getFormatter() {
    return formatter;
  }

  public void setFormatter(DataFormatter formatter) {
    this.formatter = formatter;
  }
}

File: chart/src/main/java/cn/jingzhuan/lib/chart3/data/dataset/AbstractDataSet.kt
package cn.jingzhuan.lib.chart3.data.dataset

import android.graphics.Color
import android.graphics.Rect
import cn.jingzhuan.lib.chart3.Viewport
import cn.jingzhuan.lib.chart3.axis.AxisY
import cn.jingzhuan.lib.chart3.axis.AxisY.AxisDependency
import cn.jingzhuan.lib.chart3.data.value.AbstractValue
import cn.jingzhuan.lib.chart3.formatter.IValueFormatter
import kotlin.math.roundToInt

/**
 * @since 2023-09-05
 * created by lei
 */
abstract class AbstractDataSet<T : AbstractValue> : IDataSet {

    var values: MutableList<T> = ArrayList()

    /**
     * 是否显示
     */
    var isVisible = true

    var enable = true

    var viewportYMin = Float.MAX_VALUE

    var viewportYMax = -Float.MAX_VALUE

    var minValueOffsetPercent = 0f

    var maxValueOffsetPercent = 0f

    var startXOffset = 0f

    var endXOffset = 0f

    @get:AxisDependency
    var axisDependency = AxisY.DEPENDENCY_LEFT

    var color = Color.GRAY

    var colorDynamic = false

    /**
     * 当前绘制索引
     */
    var drawIndex = -1

    var minValueCount = -1

    /**
     * 当前Dataset的标签
     */
    var tag: String? = null

    /**
     * 数据格式化器
     */
    var formatter: IValueFormatter? = null

    var maxIndex = 0

    var minIndex = 0

    var forceValueCount = -1

    constructor()

    constructor(tag: String?) {
        this.tag = tag
    }

    constructor(values: List<T>) {
        this.values = values.toMutableList()
    }

    constructor(values: List<T>, @AxisDependency axisDependency: Int) {
        this.values = values.toMutableList()
        this.axisDependency = axisDependency
    }

    override fun calcMinMax(viewport: Viewport) {}

    override fun calcMinMax(viewport: Viewport, content: Rect, max: Float, min: Float) {
        calcMinMax(viewport)
    }

    abstract fun addEntry(value: T?): Boolean

    abstract fun removeEntry(value: T?): Boolean

    fun getEntryIndex(value: T): Int {
        return values.indexOf(value)
    }

    fun getEntryForIndex(index: Int): T? {
        return values.getOrNull(index)
    }

    fun getVisiblePoints(viewport: Viewport): List<T>? {
        val allValue = ArrayList(values)
        val listSize = allValue.size
        val from = (viewport.left * listSize).roundToInt()
        val to = (viewport.right * listSize).roundToInt()
        return safeSubList(allValue, from, to)
    }

    private fun safeSubList(list: List<T>?, from: Int, to: Int): List<T>? {
        if (list.isNullOrEmpty()) return list
        val size = list.size
        val safeFrom = if (from >= size) 0 else from
        val safeTo = to.coerceAtMost(size)
        return list.subList(safeFrom, safeTo)
    }

    fun getVisibleRange(viewport: Viewport): Float {
        return (viewport.right - viewport.left) * getEntryCount()
    }
}
